rust-runtime/aws-smithy-types/src/auth.rs

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+//! Smithy HTTP Auth Types
+
+use std::cmp::PartialEq;
+use std::fmt::Debug;
+
+#[derive(Debug, PartialEq)]
+enum AuthErrorKind {
+    InvalidLocation,
+    SchemeNotAllowed,
+}
+
+/// Error for Smithy authentication
+#[derive(Debug)]
+pub struct AuthError {
+    kind: AuthErrorKind,
+}
+
+impl std::fmt::Display for AuthError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use AuthErrorKind::*;
+        match &self.kind {
+            InvalidLocation => write!(f, "invalid location: expected `header` or `query`"),
+            SchemeNotAllowed => write!(
+                f,
+                "scheme only allowed when it is set into the `Authorization header`"
+            ),
+        }
+    }
+}
+
+impl From<AuthErrorKind> for AuthError {
+    fn from(kind: AuthErrorKind) -> Self {
+        Self { kind }
+    }
+}
+
+/// Authentication configuration to connect to a Smithy Service
+#[derive(Clone, Debug, PartialEq)]
+pub struct AuthApiKey {
+    api_key: String,
+}
+
+impl AuthApiKey {
+    /// Constructs a new API key.
+    pub fn new(api_key: impl Into<String>) -> Self {
+        Self {
+            api_key: api_key.into(),
+        }
+    }
+
+    /// Returns the underlying api key.
+    pub fn api_key(&self) -> &str {
+        &self.api_key
+    }
+
+    /// Sets the value for the api key
+    pub fn set_api_key(mut self, api_key: impl Into<String>) -> Self {
+        self.api_key = api_key.into();
+        self
+    }
+}
+
+impl From<&str> for AuthApiKey {
+    fn from(api_key: &str) -> Self {
+        Self {
+            api_key: api_key.to_owned(),
+        }
+    }
+}
+
+impl From<String> for AuthApiKey {
+    fn from(api_key: String) -> Self {
+        Self { api_key }
+    }
+}
+
+/// An HTTP-specific authentication scheme that sends an arbitrary
+/// auth value in a header or query string parameter.
+// As described in the Smithy documentation:
+// https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/resources/software/amazon/smithy/model/loader/prelude.smithy
+#[derive(Clone, Debug, PartialEq)]
+pub struct HttpAuthDefinition {
+    /// Defines the location of where the Auth is serialized. This value
+    /// can be set to `"header"` or `"query"`.
+    location: String,
+
+    /// Defines the name of the HTTP header or query string parameter
+    /// that contains the Auth.
+    name: String,
+
+    /// Defines the security scheme to use on the `Authorization` header value.
+    /// This can only be set if the "in" property is set to `"header"`.
+    scheme: Option<String>,
+}
+
+impl HttpAuthDefinition {
+    /// Constructs a new HTTP auth definition.
+    fn new<S>(location: String, name: String, scheme: S) -> Result<Self, AuthError>
```

Comment:
We should use the builder pattern so that more fields can be added in the future without a breaking change.